runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -8114,8 +8114,91 @@ done:;
 	VMINLINE VM_BytecodeAction
 	withfield(REGISTER_ARGS_LIST)
 	{
+retry:
 		VM_BytecodeAction rc = EXECUTE_BYTECODE;
-		// TODO: Implement bytecode
+		const U_16 index = *(U_16 *)(_pc + 1);
+		J9ConstantPool * const ramConstantPool = J9_CP_FROM_METHOD(_literals);
+		J9RAMFieldRef * const ramFieldRef = ((J9RAMFieldRef *)ramConstantPool) + index;
+		const UDATA flags = ramFieldRef->flags;
+		const bool isVolatile = (0 != (flags & J9AccVolatile));
+		const UDATA valueOffset = ramFieldRef->valueOffset;
+		const UDATA newValueOffset = valueOffset + J9_OBJECT_HEADER_SIZE;
+		j9object_t originalObjectRef = *(j9object_t *)(_sp + (J9_ARE_ALL_BITS_SET(flags, J9FieldSizeDouble) ? 2 : 1));
+		J9Class *objectRefClass = NULL;
+		j9object_t copyObjectRef = NULL;
+
+		/* In a resolved field, flags will have the J9FieldFlagResolved bit set, thus
+		 * having a higher value than any valid valueOffset.
+		 *
+		 * This check avoids the need for a barrier, as it will only succeed if flags
+		 * and valueOffset have both been updated. It is crucial that we do not treat
+		 * a field ref as resolved if only one of the two values has been set (by
+		 * another thread that is in the middle of a resolve).
+		 */
+		if (J9_UNEXPECTED((flags <= valueOffset) || J9_ARE_NO_BITS_SET(flags, J9FieldFlagPutResolved))) {
+			/* Field is unresolved */
+			J9Method *method = _literals;
+			buildGenericSpecialStackFrame(REGISTER_ARGS, 0);
+			updateVMStruct(REGISTER_ARGS);
+			resolveInstanceFieldRef(_currentThread, method, ramConstantPool, index, J9_RESOLVE_FLAG_RUNTIME_RESOLVE | J9_RESOLVE_FLAG_FIELD_SETTER | J9_RESOLVE_FLAG_FIELD_WITHER, NULL);
+			VMStructHasBeenUpdated(REGISTER_ARGS);
+			if (immediateAsyncPending()) {
+				rc = GOTO_ASYNC_CHECK;
+				goto done;
+			} else if (VM_VMHelpers::exceptionPending(_currentThread)) {
+				rc = GOTO_THROW_CURRENT_EXCEPTION;
+				goto done;
+			}
+			restoreGenericSpecialStackFrame(REGISTER_ARGS);
+			goto retry;
+		}
+
+		/* Must be done after resolution as we don't know where the object ref is
+		 * until we have resolved the fieldref
+		 */
+		if (NULL == originalObjectRef) {
+			rc = THROW_NPE;
+			goto done;
+		}
+		objectRefClass = J9OBJECT_CLAZZ(_currentThread, originalObjectRef);
+
+		if (!J9_IS_J9CLASS_VALUETYPE(objectRefClass)) {
```

Comment:
Can this be ensure by verification instead?

Reply:
The verifier will only see the signatures. An 'L' signature may be a VT or just a regular Ref type so we cant do the check there.